File: bench/shootout.jl
# Cross-engine reader shootout, CSV.jl side. Generates the shape files once,
# times CSV.File(path) best-of-N, then runs bench/shootout.py for the others.
#   julia --project=test -t8 bench/shootout.jl <datadir> [sizeMiB] [reps]
using CSV, Tables, Printf
include(joinpath(@__DIR__, "bench_matrix.jl"))
datadir = isempty(ARGS) ? mktempdir() : ARGS[1]
mb = length(ARGS) >= 2 ? parse(Float64, ARGS[2]) : 200.0
reps = length(ARGS) >= 3 ? parse(Int, ARGS[3]) : 5
mkpath(datadir)
shapes = (:numeric, :mixed, :strings, :temporal, :quoted, :wide)
for shape in shapes
    path = joinpath(datadir, "$shape.csv")
    isfile(path) || write(path, makedata(shape, round(Int, mb * 2^20)))
end
out = open(joinpath(datadir, "shootout.tsv"), "a")
for nt in (Threads.nthreads(), 1)
    for shape in shapes
        path = joinpath(datadir, "$shape.csv")
        nbytes = filesize(path)
        f = () -> CSV.File(path; ntasks=nt, on_error=:collect)
        f()
        t = minimum((GC.gc(); @elapsed f()) for _ in 1:reps)
        @printf("%-8s %-12s %dT %9.1f ms %8.0f MiB/s\n", "CSV.jl", shape, nt, t * 1e3, nbytes / 2^20 / t)
        println(out, join(("CSV.jl", shape, nt, round(t * 1e3, digits=2), round(nbytes / 2^20 / t, digits=1)), '\t'))
        flush(out); flush(stdout)
    end
end
close(out)
py = joinpath(@__DIR__, "shootout.py")
for nt in (Threads.nthreads(), 1)
    run(`python3 $py $datadir $nt $reps`)
end
println("results: ", joinpath(datadir, "shootout.tsv"))

File: bench/shootout.py
#!/usr/bin/env python3
# Cross-engine reader shootout: polars, duckdb, pyarrow on the files that
# bench/shootout.jl generates (same bytes as bench_matrix.jl shapes).
#   python3 bench/shootout.py <datadir> <threads> [reps]
# Appends TSV rows (engine, shape, threads, ms, MiB/s) to <datadir>/shootout.tsv.
import os, sys, time, glob
datadir, threads = sys.argv[1], int(sys.argv[2])
reps = int(sys.argv[3]) if len(sys.argv) > 3 else 5
os.environ["POLARS_MAX_THREADS"] = str(threads)
import polars as pl, duckdb, pyarrow as pa, pyarrow.csv as pacsv
pa.set_cpu_count(threads); pa.set_io_thread_count(threads)
con = duckdb.connect(); con.execute(f"SET threads={threads}")
def best(f):
    f()
    return min(timeit(f) for _ in range(reps))
def timeit(f):
    t = time.perf_counter(); f(); return time.perf_counter() - t
out = open(os.path.join(datadir, "shootout.tsv"), "a")
for path in sorted(glob.glob(os.path.join(datadir, "*.csv"))):
    shape = os.path.basename(path)[:-4]
    nbytes = os.path.getsize(path)
    engines = {
        "polars":  lambda: pl.read_csv(path, try_parse_dates=True, infer_schema_length=10000),
        "duckdb":  lambda: con.execute("SELECT * FROM read_csv_auto(?)", [path]).fetch_arrow_table(),
        "pyarrow": lambda: pacsv.read_csv(path),
    }
    for name, f in engines.items():
        try:
            table = f()
            print(f"{name:8s} {shape:12s} schema={table.schema}", flush=True)
            t = best(f)
            print(f"{name:8s} {shape:12s} {threads}T {t*1e3:9.1f} ms {nbytes/2**20/t:8.0f} MiB/s", flush=True)
            out.write(f"{name}\t{shape}\t{threads}\t{t*1e3:.2f}\t{nbytes/2**20/t:.1f}\n"); out.flush()
        except Exception as e:
            print(f"{name:8s} {shape:12s} ERROR {e}", flush=True)
